### ISSUE TITLE
mensural timestamps

### ISSUE THREAD

**Issue by craigsapp** [music-encoding/music-encoding#559]
For mensural notation, how do `@tstamp`s work?   In other words do `<barLine>` events in the layer restart the timestamp calculations, or do timestamps only get reset at the start of the parent container of a staff/layer (so timestamp would start at a section that then contains a staff/layer).  Also, the time units for `@tstamp` is the bottom of a time signature.  What happens if there is no time signature?  Presumably a default of quarter notes should be used.  Mensural music typically uses the whole-note (semibreve) as the beat, so ideally this should be specified as the unit of `@tstamp` without implying it is the bottom of a time signature.

And related to the above questions, are timestamps allowed in mensural music?  For example mensural music has a complex interaction between `@dur` and `@dur.ges` in a similar way that CMN music has with `@accid` and `@accid.ges`.  Or to put it another way, the duration of a note in mensural notation is contextual (defined by both the mensuration, and the other rhythms around it as well as the dots of division).

And a third related question is how to deal with aligning multiple layers in mensural music.  If there is a divisi in the score for one chord, the entire musical part needs a second layer with an enormous number of `<space>` elements?  Alternatively, the divisi would be put into a gratutious independent `<section>`?  And what sort of durations should be in the `<space>` element?

**Comment by pe-ro** [comment]
@craigsapp, All very good questions.  I don't feel qualified to answer them, so I hope we'll both get some help from others. 😀 

Some observations --
- if there's no time signature, one may be assigned (probably also made invisible) for calculating alignment, but we may need to work out exactly how this should be done ;
- `@dur` on `<space>` can take multiple values (as it can in several other places too), which should be added together to ascertain the total duration of the "gap";
- it may be necessary to define `@tstamp` as an offset since the beginning of the file (or some other arbitrary point, such as the beginning of the `<section>`) for mensural notation;
- these are all problems inherent in mensural notation itself, not in MEI; it's why we don't use it anymore (except for historical purposes). 😁 

Hope this helps.

**Comment by craigsapp** [comment]
> if there's no time signature, one may be assigned (probably also made invisible) for calculating alignment, but we may need to work out exactly how this should be done ;

Yes, this would be good.  I have a related issue for verovio which has some discussions of proposals: 
 https://github.com/rism-ch/verovio/issues/853

> `@dur` on `<space>` can take multiple values (as it can in several other places too), which should be added together to ascertain the total duration of the "gap";

Meaning that a 5 breve rest would be:?

```
<space dur="breve breve breve breve breve">
```

But a general problem would be that breves in mensural music have an indeterminate duration.  Basically they can be imperfect, consisting of two semi-breves, or perfect, consisting of two semi-breves.  (so things can quickly get messy when trying to align mensural data based on visual durations rather than gestural ones, but gestural ones will also have a problem).

> it may be necessary to define `@tstamp` as an offset since the beginning of the file

That is interesting.  I do that sort of thing often for music analysis purpose to measure the distance between any two points in the score in quarter notes, for example.

>  (or some other arbitrary point, such as the beginning of the `<section>`) for mensural notation;

In general this will be better than from the beginning of the file, since there will be complications with `<mdiv>` and such.  Having the tstamps start at the beginning of a `<section>` seems a little more robust.  The complication is that that causes a `<section>` to act like a gigantic measure, so if adding a note to the start of a section, there would be a lot of tstamps to update (This is similar to CMN `<measure>` tstamps, but measures are typically shorter).

> these are all problems inherent in mensural notation itself, not in MEI; it's why we don't use it anymore (except for historical purposes). 😁

:smile:   A particular problem is that the `@dur.ges` of a note is not a constant value when the mensuration is changed, so relying on it is just as problematic as using `@dur` unless you don't change the mensuration signs in the score.

**Comment by pe-ro** [comment]
```
<space dur="breve breve breve breve breve">
```
Sorry, I gave you incorrect information.  `@dur` currently only allows multiple values on `<mSpace>` and control events; that is, those elements that cut across the measure hierarchy, like `<dir>`, `<slur>`, etc.

We could consider allowing `@dur` to take multiple values on `<space>` too, or in all cases, or maybe just when the mensural module is invoked, but using `@dur.ppq` is likely to be more useful than either `@dur` or `@dur.ges`. It expresses duration in terms of pulses-per-quarter. The number of pulses-per-quarter can be set for the entire score or independently for each staff.  Now, if we add `@tstamp.ppq`, we can capture a time stamp in terms of elapsed pulses. And, if we add `@tstamp.ppq.origin` (or `@tstamp.ppq.since` or some such), then one can provide the point from which `@tstamp.ppq` is measured.  If `@tstamp.ppq.origin` isn't specified, the default would be since the beginning of the score.  Does this seem useful?

Your comment broke off abruptly.  "Unless", what?

**Comment by pe-ro** [comment]
@craigsapp, As an added bonus, you'll finally have your context-less time stamp. 😀

**Comment by craigsapp** [comment]
> @craigsapp, As an added bonus, you'll finally have your context-less time stamp. 😀

🎉 🎈

**Comment by craigsapp** [comment]
> We could consider allowing `@dur` to take multiple values on `<space>` too, or in all cases, or maybe just when the mensural module is invoked, 

There is some advantage to that.  Humdrum allows partial layers, while MEI requires complete layers (or at least padding at the start of a measure to align notes in the layer with other layers).  So I have to implement automatic splitting of the implicit padding `<spaces>` when converting to MEI.  Allowing a list of durations would mostly avoid that problem and allow one padding `<space>` where I now may need multiple spaces.

A problem with allowing multiple durations is that the `<space>` would be more difficult to turn into a `<rest>`.  Spaces are invisible rests 😜 , and it is easy to convert into a visible rest since it has a duration that matches `rest@dur`.  Adding a list of durations to `<space>` would break that relation (but it is not a problem for me).

**Comment by craigsapp** [comment]
> Your comment broke off abruptly. "Unless", what?

I got distracted by something...  Probably it was supposed to be something like "unless you don't change the mensuration signs in the score".

**Comment by pe-ro** [comment]
This may be addressed in v. 4.1.0, but I'm assigning it to v. 5.0.0 for now because I can't be certain whether we'll need to break backward compatibility.
